sale_order_project/models/sale.py

```diff
@@ -22,48 +22,45 @@
 #
 ###############################################################################
 
-from openerp.osv import fields, orm
+from openerp import api, fields, models
 from datetime import date
 
 
-class sale_order(orm.Model):
+class SaleOrder(models.Model):
     _inherit = "sale.order"
 
-    _columns = {
-        'related_project_id': fields.many2one(
-            'project.project',
-            'Project',
-            readonly=True,
-            states={'draft': [('readonly', False)]}
-        ),
-    }
+    @api.one
+    @api.depends('project_id')
+    def _compute_related_project_id(self):
+        if self.project_id.use_tasks:
+            related_project_id = self.env['project.project'].search(
+                [('analytic_account_id', '=', self.project_id.id)], limit=1)
+        else:
+            related_project_id = False
+        self.related_project_id = related_project_id and related_project_id[0]
 
-    def onchange_related_project_id(self, cr, uid, ids, related_project_id, context=None):
-        project_obj = self.pool['project.project']
-        if related_project_id:
-            project = project_obj.browse(cr, uid, related_project_id, context=context)
-            return {'value': {'project_id': project.analytic_account_id.id}}
-        return {}
+    related_project_id = fields.Many2one(
+        comodel_name='project.project', string='Project',
+        compute='_compute_related_project_id')
```

Comment:
I made that change because @pedrobaeza suggested

The button continues there, so you are able to create a project from the sale.order
If you add an analytic account with use_tasks checked it will add the related project.